larq_compute_engine/mlir/transforms/fuse_padding.cc

```diff
@@ -0,0 +1,70 @@
+//#include "larq_compute_engine/core/types.h"
+//#include "larq_compute_engine/mlir/ir/lce_ops.h"
+//#include "larq_compute_engine/mlir/transforms/passes.h"
+//#include "mlir/Dialect/StandardOps/IR/Ops.h"
+//#include "mlir/IR/PatternMatch.h"
+//#include "tensorflow/compiler/mlir/lite/transforms/dilated_conv.h"
+//#include "tensorflow/compiler/mlir/lite/utils/validators.h"
+//#include "tensorflow/compiler/mlir/tensorflow/ir/tf_ops.h"
+
+#include "larq_compute_engine/mlir/transforms/padding.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "tensorflow/compiler/mlir/lite/ir/tfl_ops.h"
+
+namespace mlir {
+namespace TFL {
+
+namespace {
+
+// Version used when matching a TFLite op that has `stride_height` and
+// `stride_width` as separate attributes. Due to a TableGen limitation we can't
+// pass them both into a single function call.
+bool IsSamePaddingPartial(Attribute paddings_attr, Value input, Value output,
+                          Attribute strides_attr, uint64_t dimension) {
+  if (!paddings_attr.isa<DenseElementsAttr>()) return false;
+  auto paddings = paddings_attr.dyn_cast<DenseElementsAttr>();
+  auto input_shape = input.getType().cast<RankedTensorType>().getShape();
+  auto output_shape = output.getType().cast<RankedTensorType>().getShape();
+
+  if (!strides_attr.isa<IntegerAttr>()) return false;
+  const int stride = strides_attr.cast<IntegerAttr>().getInt();
+
+  // Check that there is no padding in the batch and channel dimensions
+  return IsNoPadding(paddings, 0) &&
+         IsSamePadding1D(paddings, dimension, input_shape[dimension],
+                         output_shape[dimension], stride) &&
+         IsNoPadding(paddings, 3);
```

Comment:
A bit weird perhaps of a function `IsSamePaddingPartial` with a `dimension` argument to also check things in dimension 0 and 3. Is it possible to remove these `IsNoPadding` checks here and call them from the place where this function is called, or is that tricky because of MLIR?

Reply:
I agree that this was weird. I've changed it in [1497455](https://github.com/larq/compute-engine/pull/712/commits/14974551bc1cee02631a9564584781335b00ddda)

Reply:
Thanks, looks good now!